docs/source/design/dataset_argument.rst

```diff
@@ -0,0 +1,58 @@
+.. -*- mode: rst -*-
+.. vi: set ft=rst sts=4 ts=4 sw=4 et tw=79:
+
+.. _chap_design_dataset_argument:
+
+********************
+``dataset`` argument
+********************
+
+All commands which operate on datasets have a ``dataset`` argument (``-d`` or
+``--dataset`` for the command line API) to identify a single dataset as the
+context of an operation.
```

Comment:
I think this section lacks a hint on what is considered the context, if no `--dataset` argument is given. And that probably is the part that makes it somewhat confusing. Context ds is either `-d`, derived from PWD or derived from a given `path` argument.

Reply:
I agree that this would be nice. But ultimately this section does not have such hint, because I am not aware of a general valid approach. The best I could come up with is "given no `dataset` declares no context, and all bets are off" i.e. the command will do whatever seems to make sense. But this is not a "design" and unknown whether actually true. Even your statement is speculation AFAICS. Let's start simple.

Reply:
Reminded me about https://github.com/datalad/datalad/issues/3905 (refers to relevant issues there as well) which unfortunately didn't get any traction, but I see this PR as a step toward addressing it.

I echo @bpoldrack 's comment that  "here" might be the correct place to state what the "context of an operation" really means, and how does it "differ" whenever `--dataset` is not specified.  E.g. `save` would add/save present but not yet "known" subdataset even without `-d` but `clone` would not unless `-d` is given.

<details>
<summary>example</summary> 

```shell
# sorry -- I start "dirty"
(git-annex)lena:/tmp/testds[master]git-annex
$> datalad status      
 modified: hcp-100206-git-annex2 (dataset)

$> datalad clone ///
install(ok): /tmp/testds/datasets.datalad.org (dataset)                                                                                                                                                              

$> datalad status   
untracked: datasets.datalad.org (directory)
 modified: hcp-100206-git-annex2 (dataset)

$> datalad clone -d. /// another.datasets.datalad.org
install(ok): another.datasets.datalad.org (dataset)                                                                                                                                                                  
add(ok): another.datasets.datalad.org (file)                                                                                                                                                                         
add(ok): .gitmodules (file)                                                                                                                                                                                          
save(ok): . (dataset)                                                                                                                                                                                                
action summary:                                                                                                                                                                                                      
  add (ok: 2)
  install (ok: 1)
  save (ok: 1)
datalad clone -d. /// another.datasets.datalad.org  4.02s user 1.93s system 38% cpu 15.410 total

$> datalad status                                    
untracked: datasets.datalad.org (directory)
 modified: hcp-100206-git-annex2 (dataset)

$> datalad --version
datalad 0.14.6
```
</details>
and I feel that some other operations also might still have this duality in their operation depending on either `--dataset` is provided or not, while often still operating on a dataset deduced to be the "context" of operation. E.g. above with `clone` the "context" is the cloned dataset, not dataset we are currently in.

So may be smth like

```suggestion
If ``--dataset`` argument is not provided, context of an operation is operation-specific (e.g., for `clone` command -- the dataset which is being cloned), but typically - a dataset which current working directory belongs to.
In the latter case, if operation (e.g., `get`) does not find a dataset in current working directory, operation fails with an ``NoDatasetFound`` error.
```

If we see it needing to be formalized more later, we could get a dedicated section on that, but IMHO we should not just completely disregard situation whenever `--dataset` is **not** provided here.